server/aesdsocket.c

```diff
@@ -119,241 +163,336 @@ int main(int argc, char** argv)
+    else
+        *malloc_buffer = (char *)malloc(sizeof(char) * bytes_count);
+        
+    if (*malloc_buffer == NULL)
     {
+        printf("Error while allocating memmory to buffer\n");
+        syslog(LOG_ERR, "Error while allocating memmory to buffer");
         return -1;
     }
 
     // copy data including \n and update malloc buffer length
-    memcpy(malloc_buffer + malloc_buffer_len, buffer, (index == buff_len ? buff_len : index + 1));
-    malloc_buffer_len += (index == buff_len ? buff_len : index + 1);
+    memcpy(*malloc_buffer + *malloc_buffer_len, buffer, bytes_count);
+    *malloc_buffer_len += bytes_count;
 
-    int ret_len = malloc_buffer_len;
-
-    if (index < buff_len)
-    {
-        if (write(file_fd, malloc_buffer, malloc_buffer_len) < 0)
-        {
-            printf("Error: %s : Detected error while writing to the file\n", strerror(errno));
-            syslog(LOG_ERR, "Error: %s : Detected error while writing to the file\n", strerror(errno));
-            ret_len = 0;
-        }
-        free(malloc_buffer);
-        malloc_buffer = NULL;
-        malloc_buffer_len = 0;
-
-        return ret_len;
-    }
-
-    return 0;
+    if (index < buffer_len)
+        return 1;
+    else
+        return 0;
 }
 
 /**
- * @brief   The function process the data read from the client. It mallocs buff_len
- *          size and copy data from the read buffer. If '\n' is found in the buffer
- *          then entire string will be copied to the file.
- * 
- * @param   
- *  *buffer     A char array contains data read from the client
- *  buff_len    Number of bytes present in the *buffer array.
- * 
- * @return  Returns number of bytes written to the file on success. Returns -1 when 
- *          malloc fails.
-*/
-int get_write_data(char **malloc_buffer, int total_size, int file_fd)
+ * @brief   Reads all the bytes from the file and copies it to the
+ *          malloc buffer.
+ *
+ * @param   file_fd: File descriptor
+ * @param   malloc_buffer: Dynamically allocated, all bytes read
+ *          from the file are copied to this.
+ * @param   malloc_buffer_len: Updated to the size of malloc buffer.
+ *
+ * @return  Returns 0 on success else < 0 value is returned on error. 
+ */
+int file_read(int file_fd, char **malloc_buffer, int *malloc_buffer_len)
 {
-    if (*malloc_buffer != NULL)
-        free(*malloc_buffer);
+    int char_count;
+    char char_data;
+    *malloc_buffer_len = 0;
+
+    pthread_mutex_lock(&file_lock);
+    lseek(file_fd, 0, SEEK_SET);
+    for (char_count = 0; read(file_fd, &char_data, 1) > 0; char_count++);
+    pthread_mutex_unlock(&file_lock);
 
-    *malloc_buffer = (char *)malloc(sizeof(char) * total_size);
+    if (char_count <= 0)
+    {
+        printf("File is empty!\n");
+        syslog(LOG_ERR, "File is empty!");
+        return -1;
+    }
+
+    if (*malloc_buffer)
+        *malloc_buffer = (char *)realloc(*malloc_buffer, char_count);
```

Comment:
@ajaykandagal what does this assume about the size of the heap and the size of all packets sent?